### DISCUSSION TITLE
Custom Node Script

### DISCUSSION THREAD

**Discussion by automartin5000** [projen/projen#3464] [Q&A] [answered]
If I want to create a custom TS-based script, what's the recommended way to do that? In my case, I want to create a step in a task that calls and API and stores the value in an environment variable. Normally I might do something like `ts-node script.ts` or `node script.js`, but I wasn't sure if there's a better Projen way. For example, I imagine the `ts-node` way wouldn't work with non-Typescript projects.

**Comment by automartin5000** [comment]
Am I maybe looking for "[3rd party builtin tasks](https://github.com/projen/projen/issues/2051#issuecomment-1320047876)"?

**Reply by mrgrain** [reply]
Yes I think that hypothetical feature would make it easier to do something like this. Although the specific RFC comment was more directed at 3rd party libraries providing re-usable scripts, since they need to somehow ship with the script code and that makes it slightly more difficult.

**Comment by mrgrain** [comment, marked as answer]
There is no standard way to do this at the moment. However it's also not very difficult.

Here's an example for a custom script step:
https://github.com/cdklabs/awscdk-service-spec/blob/main/projenrc/download-script.ts

This can than be used like this:
```ts
project.addTask('task-name', {
  steps: [new DownloadScript('url', 'target', true)],
});
```
